### PR TITLE
o/snapshotstate: set snapshot set id from its filename

### PR DESCRIPTION
**Pull request by stolowski**
Set set id from snapshot filename, which overrides set id read from meta. This change is annoying to unit-test currently, so I've provided spread check.

Also remove part of the iteration check landed previously as it won't be needed after this change.

### CODE REVIEW

File: overlord/snapshotstate/backend/backend.go [hunk 1]
```diff
@@ -179,6 +172,7 @@ func Filename(snapshot *client.Snapshot) string {
 }
 
 func isSnapshotFilename(fname string) (ok bool, setID uint64) {
```
Comment (mvo5) [inline overlord/snapshotstate/backend/backend.go:174]: (nitpick) I would probably rename the function input far from "fname" to something like "path" or "filePath" or "fullPath" or something. Unfortunately we don't really have a convention for this, i.e. when fname or fn means just the filename or the filename with path.

File: overlord/snapshotstate/backend/reader.go [hunk 2]
```diff
@@ -84,6 +84,15 @@ func Open(fn string) (reader *Reader, e error) {
 		return nil, err
 	}
 
+	// XXX: this mirrors the check from Iter()
+	ok, setID := isSnapshotFilename(fn)
+	if !ok {
+		return nil, fmt.Errorf("not a snapshot filename: %q", fn)
+	}
+	// set id from the filename has the authority and overrides the one from
```
Comment (mvo5) [inline overlord/snapshotstate/backend/reader.go:92]: Thanks for this comment.

**Review by mvo5 — APPROVED** [review]
Thank you, I think that is good.

File: tests/main/snapshot-basic/task.yaml [hunk 3]
```diff
@@ -125,3 +125,14 @@ execute: |
...
     test "$(cat memory-kb.txt)" -lt 40000
+
+    # check that snapshot set id from the filename has authority
+    snap install test-snapd-sh
+    snap save test-snapd-sh
+    SNAPSHOT_FILE=$(ls /var/lib/snapd/snapshots/*test-snapd-sh*.zip)
+    NEWSNAPSHOT_FILE=$(echo "$SNAPSHOT_FILE" | sed -e's/[1-9]\+_/123_/')
+    # rename the snapshot file to force a new set id.
+    mv "$SNAPSHOT_FILE" "$NEWSNAPSHOT_FILE"
+    snap saved | MATCH "123 +test-snapd-sh"
```
Comment (bboozzoo) [inline tests/main/snapshot-basic/task.yaml:136]: Should we also verify that a snapshot is only listed once with the new set ID?

Comment (stolowski) [reply tests/main/snapshot-basic/task.yaml:136]: Added.

File: overlord/snapshotstate/backend/backend.go (hunk 1 above)
Comment (stolowski) [reply overlord/snapshotstate/backend/backend.go]: Done.

File: overlord/snapshotstate/backend/reader.go [hunk 4]
```diff
@@ -84,6 +84,15 @@ func Open(fn string) (reader *Reader, e error) {
 		return nil, err
 	}
 
+	// XXX: this mirrors the check from Iter()
```
Comment (pedronis) [inline overlord/snapshotstate/backend/reader.go:87]: why the XXX bit of this here?  either we are happy?  or we add a setID argument to Open that we use if it's not 0, which seems easy enough if making the diff a bit larger?

leaving the XXX for the future seems not very useful here

**Review by pedronis — APPROVED** [review]
@mvo, @stolowski   I really hope we can get a unit test for this in a follow up.

Also let's solve or not have the XXX

**Pull request merged**